feat: support configuring multiple catalog index images [RHIDP-12935] (#2717)
redhat-developer/rhdh-operator: docs/dynamic-plugins.md

By default, the `rhdh` profile of operator [injects](../config/profile/rhdh/patches/deployment-patch.yaml#L31-L32) the `CATALOG_INDEX_IMAGE` environment variable in the RHDH `install-dynamic-plugins` init container.
To use a different catalog index image, such as a newer version or a mirrored image, use the `extraEnvs` field in your Backstage CR. See [examples/catalog-index.yaml](../examples/catalog-index.yaml) for a complete example.

### Extra catalog index images

> In addition to the primary catalog index image, you can configure extra catalog index images using the `EXTRA_CATALOG_INDEX_IMAGES` environment variable. This allows loading plugin configurations from multiple catalog index images. See [Using extra catalog index images](https://github.com/redhat-developer/rhdh/blob/main/docs/dynamic-plugins/installing-plugins.md#using-extra-catalog-index-images) for more details on how RHDH handles this environment variables.

Author: ```suggestion
In addition to the primary catalog index image, you can configure extra catalog index images using the `EXTRA_CATALOG_INDEX_IMAGES` environment variable.
Note that extra catalog index images only allow contributing additional catalog entities to the RHDH Extensions Backend plugin, so they can be listed in the Extensions UI. Only the primary catalog index (`CATALOG_INDEX_IMAGE`) is used for default plugin configuration data (`dynamic-plugins.default.yaml). See [Using extra catalog index images](https://github.com/redhat-developer/rhdh/blob/main/docs/dynamic-plugins/installing-plugins.md#using-extra-catalog-index-images) for more details on how RHDH handles this environment variable.
```

By default, the `rhdh` profile of operator [injects](../config/profile/rhdh/patches/deployment-patch.yaml#L31-L32) the `CATALOG_INDEX_IMAGE` environment variable in the RHDH `install-dynamic-plugins` init container.
To use a different catalog index image, such as a newer version or a mirrored image, use the `extraEnvs` field in your Backstage CR. See [examples/catalog-index.yaml](../examples/catalog-index.yaml) for a complete example.

> ### Extra Catalog Index Images

Author: ```suggestion
### Extra catalog index images
```

> In addition to the primary catalog index image, you can configure extra catalog index images using the `EXTRA_CATALOG_INDEX_IMAGES` environment variable. This allows loading plugin configurations from multiple catalog index images.

Author: ```suggestion
In addition to the primary catalog index image, you can configure extra catalog index images using the `EXTRA_CATALOG_INDEX_IMAGES` environment variable. This allows loading plugin configurations from multiple catalog index images. See [Using extra catalog index images](https://github.com/redhat-developer/rhdh/blob/main/docs/dynamic-plugins/installing-plugins.md#using-extra-catalog-index-images) for more details on how RHDH handles this environment variables.
```

The value is a comma-separated list of entries. Each entry supports two forms:
- **`name=image_ref`**: Assigns an explicit name to the catalog index image, which controls the extraction subdirectory under `/extensions/extra/<name>/`.
- **`image_ref`**: A direct image reference without a name; the extraction directory is auto-generated from the image reference.

To configure extra catalog index images, use the `extraEnvs` field in your Backstage CR:

```yaml
apiVersion: rhdh.redhat.com/v1alpha6
kind: Backstage
metadata:
  name: my-backstage
spec:
  application:
    extraEnvs:
      envs:
        - name: EXTRA_CATALOG_INDEX_IMAGES

>           value: "community=quay.io/rhdh/extra-index-community:1.0,internal=registry.example.com/catalog:latest"

Author: ```suggestion
          value: "rhdh-community=quay.io/rhdh-community/plugin-catalog-index:1.10,internal=registry.example.com/rhdh-catalog:latest"
```

For example, setting the following env vars on the operator deployment:
```yaml
- name: RELATED_IMAGE_extra_catalog_index_community
  value: "registry.example.com/rhdh/extra-index-community:1.0"
- name: RELATED_IMAGE_extra_catalog_index_internal
  value: "registry.example.com/rhdh/extra-index-internal:2.0"
```
will result in `EXTRA_CATALOG_INDEX_IMAGES=community=registry.example.com/rhdh/extra-index-community:1.0,internal=registry.example.com/rhdh/extra-index-internal:2.0` being set on the init container.

This follows the same pattern as the existing `RELATED_IMAGE_catalog_index` env var for the primary catalog index image. If a user also sets `EXTRA_CATALOG_INDEX_IMAGES` via `extraEnvs` or a deployment patch in their CR, the user-specified value takes precedence.

> 

Author: ```suggestion
```

The value is a comma-separated list of entries. Each entry supports two forms:
- **`name=image_ref`**: Assigns an explicit name to the catalog index image, which controls the extraction subdirectory under `/extensions/extra/<name>/`.
- **`image_ref`**: A direct image reference without a name; the extraction directory is auto-generated from the image reference.

To configure extra catalog index images, use the `extraEnvs` field in your Backstage CR:

```yaml
apiVersion: rhdh.redhat.com/v1alpha6
kind: Backstage
metadata:
  name: my-backstage
spec:
  application:
    extraEnvs:
      envs:
        - name: EXTRA_CATALOG_INDEX_IMAGES
          value: "rhdh-community=quay.io/rhdh-community/plugin-catalog-index:1.10,internal=registry.example.com/rhdh-catalog:latest"
          containers:
            - install-dynamic-plugins
```

> Alternatively, the operator can automatically populate `EXTRA_CATALOG_INDEX_IMAGES` from environment variables prefixed with `RELATED_IMAGE_extra_catalog_index_`. This is useful for injecting extra catalog index images at the operator level (e.g., via the operator deployment or subscription), without requiring changes to individual Backstage CRs.

Author: ```suggestion
Alternatively, the operator can automatically populate `EXTRA_CATALOG_INDEX_IMAGES` in operands from environment variables prefixed with `RELATED_IMAGE_extra_catalog_index_` and defined in the controller environment. This is useful for injecting extra catalog index images at the operator level (e.g., via the operator deployment or subscription), without requiring changes to individual Backstage CRs.
This is also useful for airgap tools that need to mirror all operator related images based on what is declared in the operator bundle manifest.
```

> Each environment variable of the form `RELATED_IMAGE_extra_catalog_index_<name>=<image_ref>` contributes an entry `<name>=<image_ref>` to the comma-separated `EXTRA_CATALOG_INDEX_IMAGES` value. For example, setting:

Author: ```suggestion
Each environment variable of the form `RELATED_IMAGE_extra_catalog_index_<name>=<image_ref>` in the controller contributes an entry `<name>=<image_ref>` to the comma-separated `EXTRA_CATALOG_INDEX_IMAGES` value in the RHDH operand (`install-dynamic-plugins` init container only). For example, setting these in the operator:
```

```
RELATED_IMAGE_extra_catalog_index_community=quay.io/rhdh-community/plugin-catalog-index:1.10
RELATED_IMAGE_extra_catalog_index_internal=registry.example.com/rhdh-catalog:latest
```

> results in `EXTRA_CATALOG_INDEX_IMAGES` being set to `community=quay.io/rhdh-community/plugin-catalog-index:1.10,internal=registry.example.com/rhdh-catalog:latest` (or in reverse order).

Author: ```suggestion
results in `EXTRA_CATALOG_INDEX_IMAGES` being set to `community=quay.io/rhdh-community/plugin-catalog-index:1.10,internal=registry.example.com/rhdh-catalog:latest` (or in reverse order) in any RHDH operand's `install-dynamic-plugins` init container (unless explicitly overridden in the Backstage CR).
```

### Extra catalog index images

In addition to the primary catalog index image, you can configure extra catalog index images using the `EXTRA_CATALOG_INDEX_IMAGES` environment variable. This allows loading plugin configurations from multiple catalog index images. See [Using extra catalog index images](https://github.com/redhat-developer/rhdh/blob/main/docs/dynamic-plugins/installing-plugins.md#using-extra-catalog-index-images) for more details on how RHDH handles this environment variables.

The value is a comma-separated list of entries. Each entry supports two forms:
- **`name=image_ref`**: Assigns an explicit name to the catalog index image, which controls the extraction subdirectory under `/extensions/extra/<name>/`.
- **`image_ref`**: A direct image reference without a name; the extraction directory is auto-generated from the image reference.

To configure extra catalog index images, use the `extraEnvs` field in your Backstage CR:

```yaml
apiVersion: rhdh.redhat.com/v1alpha6
kind: Backstage
metadata:
  name: my-backstage
spec:
  application:
    extraEnvs:
      envs:
        - name: EXTRA_CATALOG_INDEX_IMAGES

>           value: "rhdh-community=quay.io/rhdh-community/plugin-catalog-index:1.10,internal=registry.example.com/rhdh-catalog:latest"

Author: ```suggestion
          value: "rhdh-community=quay.io/rhdh-community/plugin-catalog-index:1.10,registry.example.com/rhdh-catalog:latest"
```
